### ISSUE TITLE
(learnyounode #6 - Make it Modular) Can someone explain to me why my code is acting weird?

### ISSUE THREAD

**Issue by pavelcreates** [nodeschool/discussions#1404]
So, I thought I was on the right track... where did I go wrong?

**mymodule.js**

``` javascript
function myModule(file, filter, callback) {

    var fs = require('fs'),
        path = require('path'),
        n = 0,
        results = [];

    fs.readdir(file, function(err, data) {
        if (err) {
            return callback(err);
        }

        for (i = 0; i < data.length; i++) {

            if (path.extname(data[i]) === "." + filter) {
                results[n] = data[i];
                n++;
            }

            callback(null, results);
        }

    });

}

module.exports = myModule;
```

**makeitmodular.js**

``` javascript
var myModule = require('./mymodule'),
    file = process.argv[2],
    filter = process.argv[3];

myModule(file, filter, function(err, result) {
    for (var i = 0; i < result.length; i++) {
        console.log(result[i]);
    }
})
```

My results come out as such:

```
CHANGELOG.md
CHANGELOG.md
LICENCE.md
CHANGELOG.md
LICENCE.md
README.md
CHANGELOG.md
LICENCE.md
README.md
CHANGELOG.md
LICENCE.md
README.md
CHANGELOG.md
LICENCE.md
README.md
CHANGELOG.md
LICENCE.md
README.md
// continues for a while in the same pattern - I shortened it to reduce the spam
```

**Comment by awong650** [comment]
haha.  I had this exact same problem.  I will give you a hint and if it isn't enough, just post another message.  Do you ever wonder what happens to variables in for statements?  If you have two for statements with the same variable (i), what do you think happens to that variable when the program runs?

**Comment by pavelcreates** [comment]
No dice...

I changed my **makeitmodular.js** file as such:

```javascript
var myModule = require('./mymodule'),
	file = process.argv[2],
	filter = process.argv[3];

myModule(file, filter, function(err, result) {
	for (var x = 0; x < result.length; x++) {
		console.log(result[x]);
	}
})
```

**Comment by awong650** [comment]
Well, one problem has been fixed.  Now you have an async problem.  Your console.log is now running at the same time as your myModule function and causing a new issue, while your module is populating your list, your main program is printing out results!

**Comment by GiannisGl** [comment]
Hello.
I don't know if you fixed it, since it is quite a long time before, but I just started learning and I saw that this is still open. 
The reason awong650 said is false. The problem is the position of your callback(null, results); which is in the for loop, so excecutes for every loop. You want to take it after the loop, so that it calls callback once for the final result. Then it should work fine.
Cheers
